Add precision tests for CUDA half linspace+logspace

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31962 Add precision tests for CUDA half linspace+logspace**

I added precision tests for CUDA half, float, and double.

The precision for CUDA half seems bad, but I checked the numbers against
previous versions of pytorch. The output of CUDA Half linspace+logspace
are exactly the same when compared with 1.2.0.

Test Plan:
- Run CI

Differential Revision: [D19320182](https://our.internmc.facebook.com/intern/diff/D19320182)

zou3519 committed:
Add precision tests for CUDA half linspace+logspace

I added precision tests for CUDA half, float, and double.

The precision for CUDA half seems bad, but I checked the numbers against
previous versions of pytorch. The output of CUDA Half linspace+logspace
are exactly the same when compared with 1.2.0.

Test Plan:
- Run CI

[ghstack-poisoned]

gchanan commented on test/test_torch.py:
```diff
@@ -14117,24 +14117,70 @@ def test_cat_big(self, device):
+# Since linspace/logspace are deterministic, we can compute an expected 
+# amount of error (by testing without a precision override), adding a tiny
+# amount (EPS) to that, and using that value as the override.
+EPS = 1e-5
```
can you put this inside the class and call it SPACE_EXTRA_EPS or something?

gchanan replied:
or maybe LINLOG_EXTRA_EPS?

zou3519 replied:
I'll rename it to LINSPACE_LOGSPACE_EXTRA_EPS.

We can't really put the variable into the class. `self` is not available when the decorator resolves, and `TestDevicePrecision` doesn't exist because our test framework modifies it to be `TestDevicePrecisionCUDA`. Using `TestDevicePrecisionCUDA.LINSPACE_LOGSPACE_EXTRA_EPS` is a bad idea because this code could be run from a machine that has a non-CUDA device and no CUDA devices (I suspect it runs on XLA right now but I'm not sure).

zou3519 committed:
Update on "Add precision tests for CUDA half linspace+logspace"


I added precision tests for CUDA half, float, and double.

The precision for CUDA half seems bad, but I checked the numbers against
previous versions of pytorch. The output of CUDA Half linspace+logspace
are exactly the same when compared with 1.2.0.

Test Plan:
- Run CI

[ghstack-poisoned]
